src/metabase/driver/postgres.clj

```diff
@@ -283,24 +283,16 @@
 
 (defn- json-query [identifier nfc-field]
   (letfn [(handle-name [x] (if (number? x) (str x) (name x)))]
-    (let [field-type           (:effective_type nfc-field)
+    (let [field-type           (:database_type nfc-field)
           nfc-path             (:nfc_path nfc-field)
           unwrapped-identifier (:form identifier)
           parent-identifier    (field/nfc-field->parent-identifier unwrapped-identifier nfc-field)
-          ;; Array and sub-JSON coerced to text
-          cast-type            (cond
-                                 (isa? field-type :type/Integer)
-                                 :type/Integer
-                                 (isa? field-type :type/Float)
-                                 :type/Float
-                                 (isa? field-type :type/Boolean)
-                                 :type/Boolean
-                                 :else
-                                 :type/Text)]
-      (hx/cast cast-type
-               (apply hsql/call [:json_extract_path_text
-                                 (hx/cast :json parent-identifier)
-                                 (mapv #(hx/cast :text (handle-name %)) (rest nfc-path))])))))
+          names                (format "{%s}" (str/join "," (map handle-name (rest nfc-path))))]
+      (reify
+        hformat/ToSql
+        (to-sql [_]
+          (hformat/to-params-default names "nfc_path")
+          (format "(%s#>> ?::text[])::%s " (hformat/to-sql parent-identifier) field-type))))))
```

Comment:
You might want to make use of the existing `pg-conversion` we've defined already for the `::` operator stuff:

https://github.com/metabase/metabase/blob/master/src/metabase/driver/postgres.clj#L271-L282

Reply:
I half-assedly tried it and it wasn't happy times with the interaction between the reify macro ToSql, but I can whole-assedly try it, sure

Reply:
For what it's worth you don't really need to reify a `ToSql` here at all. You can just have it return `(hsql/call ::json-query ...)` or something and register `::json-query` as a function handler.

Here's an example of doing that

https://github.com/metabase/metabase/blob/e273c6bf3bb77d6e189b39b68e46cedf07a4e8d4/src/metabase/models/params/chain_filter.clj#L478-L482

I know we have a lot of `ToSql`s all over the place but it's really not necessary for most of them -- part of it is my fault since I forgot that you could register your own function handlers. At any rate in HoneySQL 2 the `ToSql` protocol doesn't exist anymore so registering a function handler means the code will be easier to migrate

Reply:
I'm going to lard the fix for this in with the putative group-by fix because it's going to change it a bit